addons/purchase_stock/models/stock_move.py

```diff
@@ -62,8 +62,8 @@ def _get_price_unit(self):
             for invoice_line in line.sudo().invoice_lines:
                 if invoice_line.move_id.state != 'posted':
                     continue
-                # Discount applied on bill prior to reception
-                if invoice_line.discount and not move_layer:
+                # Discount applied on bill prior to reception of original pick or backorders
+                if invoice_line.discount and (not move_layer or move_layer.stock_move_id.picking_id[-1] == self.picking_id.backorder_id):
```

Comment:
The goal was to be sure we invoiced before receipt we should rather compare the quantities. If `invoiced_qty > receipt_qty + self.quantity` we get the discount value directly. so 
```python
price_unit = invoice_line.price_subtotal / invoice_line.quantity
```
if not, the layer will be corrected later. so 
```python
price_unit = invoice_line.price_unit
```